### ISSUE TITLE
Investigate options to speed up reconciliation up for large amount of GWs and Listeners and policies

### ISSUE THREAD

**Issue by trepel** [Kuadrant/kuadrant-operator#1085]
## Overview

Recently a scale test has been implemented:
https://github.com/Kuadrant/testsuite/tree/main/scale_test

I did run a few scale test runs. It took quite some time for Kuadrant operator to reconcile all the policies. Many AuthPolicies and RLPs did not get any status for quite some time. After they got some status it often was:
`'AuthPolicy waiting for the following components to sync: [AuthConfig (0cbc22a687a9ff2a57c54007e8ad9b6bc17de3744144196b9b8286fb1593f495)]'`
and
`'RateLimitPolicy waiting for the following components to sync: [Limitador]'`

Everything got reconciled successfully and policies got enforced eventually, but it took quite some time:
1GW 16 Listeners -> 16s to get status for all policies
1 GW 32 Listeners -> 120s to get status for all policies
1 GW 48 Listeners -> 7 min to get status for all policies
1 GW 63 Listeners -> 30 min to get status for all policies

In the operator log there were a lot of entries complaining about invalid paths. So I made the HTTPRoutes target specific Listener rather than the whole Gateway. This make the results much nicer:

1 GW 32 Listeners -> 18s to get status for all policies
1 GW 48 Listeners -> 60s to get status for all policies
1 GW 63 Listeners -> 120s to get status for all policies

I tried with 2 GWs as well:
2 GW 16 Listeners -> 18s to get status for all policies
2 GW 32 Listeners -> 76s to get status for all policies

However, this was still too much:
2 GW 63 Listeners -> 16 min to get status for all policies

## Initial Investigation

Be aware that certificate generation and DNS record creation might affect the results. It takes some for certificates to get created (the scale test uses self signed cluster issuer) and it also takes time for cloud provider to issue that many DNS records.

It seems reasonable that wasm config (only one per GW) is a contention point that said it should eventually get there. 
There are repeated log entries of “failed to update the object has been modified; please apply your changes to the latest version” in Kuadrant operator pod.

Also entries like "failed to create SOMETHING, SOMETHING already exists" appeared in Kuadrant operator pod log (SOMETHING being Certificate/AuthConfig typically) - not sure if this indicates some issue or not.

## Questions / Investigation required

Does it make sense that having many invalid paths is so expensive?
What can be done to improve on that 16 minutes? 2 Gateways and 63 Listeners on each are not super high numbers.

## Steps to reproduce

Basically follow the readme of the scale test:
https://github.com/Kuadrant/testsuite/tree/main/scale_test
I used OCP on AWS (6 worker nodes) and DNS setup on the same AWS account
This was done against Kuadrant v1.0.1, OCP V4.17.7

It is enough details here I believe, for even more detais see (Red Hat only, sorry):
https://docs.google.com/document/d/1ATH2aZJ7-qlYTV3jF_rZduMC1MTPKoWCD-N4LmmcaMA/edit?tab=t.0

**Comment by trepel** [comment]
PR with added sectionName:
https://github.com/Kuadrant/testsuite/pull/610

**Comment by trepel** [comment]
@Boomatang great news! Thanks. I will look into this - I don't want to promise today but no later than tomorrow hopefully.

**Comment by trepel** [comment]
A way to improve the time is to increase the vCPU requests/limits for Kuadrant operator pod (big thanks to @Boomatang to point at this). Currently both is set to 200m. I tried "2 Gateway 32 Listerners each scenario" with requests/limits set to 1500m and the time AuthP/RLPs waiting for .status decreased from 76s-120s to 13s. The peak vCPU consumption was sligthly over 900m during the scale test run.

Q: The fact that Kuadrant operator was throttled was not obvious to me, do we have any alerting in place or plan to have any?

Another thing to look at is the scale test itself. It might make sense to split it so that issuing Certificates and issuing dnsrecords in DNS cloud provider are done in separate step. Performance of this is not directly related to Kuadrant. Furthermore there is separate DNSOperator scale test on DNSOperator level under development:
https://github.com/Kuadrant/dns-operator/pull/326

Yet another thing is that kube-burner creates resources one by one (in relatively quick succession but not really at once). Given how our reconciliation flow logic works we might get better results if everything is created at once.

**Comment by Boomatang** [comment]
>Q: The fact that Kuadrant operator was throttled was not obvious to me, do we have any alerting in place or plan to have any?

Currently the operator will not install any alerts for the user. There are example alerts that the user can create. I don't think there is any plan for the kuadrant operator to do the installation of the alerts. Over the next while I hope to do a bit of research into where the operator was spending the cpu cycles. As it stands we don't know if that is a large load.

I agree that it might be good to split the tests out, but also we do need to be seeing how all the components play together. I think we need both .

Again with kube-burner creating resources slowly could be more common as a user adds more services into the system, but on pod restarts and upgrades it would be a bulk reconcile. So it is hard to know which is the one to more effort into.

**Comment by Boomatang** [comment]
@trepel I would like to follow up here. Is there more work required?, can regard this as being addressed?

What is the next action?

**Comment by trepel** [comment]
I believe more investigation is required. The scale test as is is still quite slow. Next step would be to run it and try to investigate further, where other bottlenecks are.

However, some changes might be done on the scale test scenario as well
- creating CRs at once (rather than one by one). This can be achieved by scaling the kuadrant operator down when the CRs are being created
- split TLSPolicy and DNSPolicy to be created in separate job. We might not need/want to have the times polluted by waiting for Certificates and/or DNS records to be issues (1) it is done by 3rd parties 2) DNS Operator has a separate scale test)
-

**Comment by Boomatang** [comment]
@trepel I have being looking in to this to understand some of the moving parts. What I am work with is not finished, but I have found enough information that I think it is worth sharing at this point. 

As the limitador CR is a singleton, I went with the assumption that would be a bottle neck, and focused on the rate limit policies. Here is the [repo](https://github.com/Boomatang/probable-octo-bassoon/tree/dev) that I was using to make the tests. It is still a work in progress  and not documented how to use it, also make sure your of the dev branch. If you want me to help getting that running, I have problem in helping.

# How it is working.
## Resource creation
The test follows the pattern of the [scale test](https://github.com/Kuadrant/testsuite/blob/main/scale_test/config.yaml) that ye use. Where a resource is created, and we wait till that resource is in an expected state before more on to the next resource. For example if a rate limit policy is created against the gateway, it waits for the accepted status to be  true, while if the policy is targeting a HTTProute it waits for the enforced status to be true. One issue with this approach is the resources need to be created on the cluster in order, see [order example](https://github.com/Boomatang/probable-octo-bassoon/blob/dev/config_sample.toml#L11)

The configuration allows the setting of [workers](https://github.com/Boomatang/probable-octo-bassoon/blob/dev/config_sample.toml#L12), which is the number of async routines that will process the resources in parallel. If no workers is specified, all the resources get its own worker, and a big bang approach is used for applying the resources.

Due to suspected  limitations of kind cluster the most policy I could create was 48 policies. These had to be split up between routes and gateways. 

## Data collection 
For collecting the data, there is a [collector](https://github.com/Boomatang/probable-octo-bassoon/blob/dev/probable_octo_bassoon/collector.py) which would get events on the resources let it be a status or spec update and save to record to a sqlite database.

There is also a timer that pings the limitador pod to get the last time the configuration was successfully update. This was also added as an entry into the sqlite database.

# Reviewing the timings
From looking over the runs there is a common pattern occurring. This timing are from when the collector was notified, and could lag behind the operator actually events. 
- policy creation on cluster to limitador CR, wasmplugin and limitador configmap being update: 3 seconds -1/+2
- limitador configmap update to limitador pod report a successful update of the internal configuration: less than 1 second, with a small few of outliers taken up to 3 seconds. These outliers only happened in one run.
- limitador pod report successful configuration update till policy status reporting Enforced as true: 5 to 10 seconds, with case taken up to 15 seconds.

What I have noticed is the rate of applying resources does not affect the timings on an individual resource. The same timing patterns were observed with 1, 2, 6 workers and all the workers.

# Changes in the system
The only change I made to the default `make local-setup` was bumping the cpu requests and limits form 200m to 4. This was done so as not to introduce a bottle neck there. I did not have a log of the cpu and memory usage for the runs. From the rough eye balling it seems the cpu needed was around 2. Or 10x what we suggest.

# Observed Issues and gaps in information
There is a number of issues that I think I was hitting in the test.
- Kind cluster limiting some how limiting the number of policies that could be created.
- The lag of applying policies to the cluster, and the api server reporting that resource creation. This was even through with applying resources to the cluster without kuadrant running. 
- The kind api server is not rate limited by default, and I have a gut feeling I may have being over loading the server. 
  - The adding of resources creates a connection to check the status in  expected state before doing the next resource
  - The collector opens connections to to get the changes to the spec and status of the resources.
  - Kuadrant its self would open a number of connections
  - I was also using k9s local to inspect the resources will the runs were going on.
- I don't record the time the resource is applied to the cluster,  only the time when the api server says it was created.
- Resources have to be created in order. While not a bad thing there might be other issues hiding in the out of order creation. 
- No defaults and overrides were used in the set up. Again this might be hiding issues.
- There was no collection of data around when a route would have being send to be protected. This would requiring pinging the routes with traffic and record when they start to be rate limited.

# Open questions
1. What is an acceptable time to enforcement? If a user creates a policy how long before the status say it is enforced?
2. Does the adding of a policy affect the protection of other previously protected routes? I expect not, but as the wasm shim and limitador are updated it would be good to prove it. Follow up question. If a malformed policy is created, how would that affect the system?
3. Is the scale test working as expected? Or in a fashion that makes the test be slow? I ask this from seeing no notice difference in applying 1 resource at time to apply all the resources at the same time. If you look at the time it took for a single policy to become enforced. Of course the rate the resources are applied to the cluster will change the runtime length, but not the  policy enforcement length.

**Comment by Boomatang** [comment]
@maleck13 ping for visibility
